### PR TITLE
Build against spark3.0.0-preview

### PR DESCRIPTION
**Pull request by WeichenXu123**
Build against spark3.0.0-preview

Fix some code which is broken under spark 3.0:
* update some breeze operator
* Update IDFModel initialization
* Update Tree node initialization

Disable some spec test which is broken, these may be due to model/transformer behavior change, fix them require much work, we could fix them in separate PR. 
* HashingTermFrequencyParitySpec
* GaussianMixtureParitySpec
* MaxAbsScalerParitySpec
* MinMaxScalerParitySpec

### CODE REVIEW

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/tree/decision/SparkNodeWrapper.scala [hunk 1]
```diff
@@ -43,7 +43,7 @@ object SparkNodeWrapper extends NodeWrapper[tree.Node] {
 
   override def leaf(node: LeafNode, withImpurities: Boolean): tree.Node = {
     val calc: ImpurityCalculator = if(withImpurities) {
-      ImpurityCalculator.getCalculator("gini", node.values.toArray)
+      ImpurityCalculator.getCalculator("gini", node.values.toArray, rawCount = -1L)
```
Comment (WeichenXu123) [inline mleap-spark/src/main/scala/org/apache/spark/ml/bundle/tree/decision/SparkNodeWrapper.scala]: Note: This `rawCount` argument is added in spark 3.0.0. Though, in mleap supporting this argument require a big change. We need to modify tree node proto definition and related serializer in mleap.
So here I simply set it to be `-1` to represent this property is unsupported in mleap. This `rawCount` in spark is just an extra statistic property for tree node and won't be used in model prediction. So leave it empty (-1) will be safe.
We could do the work of completely supporting `rawCount` for mleap  in separate PR

File: mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/IDFOp.scala [hunk 2]
```diff
@@ -25,13 +25,18 @@ class IDFOp extends SimpleSparkOp[IDFModel] {
 
     override def load(model: Model)
                      (implicit context: BundleContext[SparkBundleContext]): IDFModel = {
-      val idfModel = new feature.IDFModel(Vectors.dense(model.value("idf").getTensor[Double].toArray))
+      val idf = Vectors.dense(model.value("idf").getTensor[Double].toArray)
+      val idfModel = new feature.IDFModel(idf, new Array[Long](idf.size),0)
```
Comment (WeichenXu123) [inline mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/IDFOp.scala]: Note:
In spark 3.0, `org.apache.spark.mllib.feature.IDFModel` (note is in mllib package) take 3 arguments now. Make mleap support these new added 2 arguments (`docFreq` and `numDocs`) require a big change, we need to modify `ml.combust.mleap.core.feature.IDFModel` and `IDFOp` in mleap. We could leave the work in separate PR.
Here I fill this two arguments with zeros, this keep the same behavior with spark IDFModel reader (when reader model saved by old version spark).

**Comment by WeichenXu123** [comment]
Note:
my local test passed. my command is
`sbt ++2.12.10 test` (do not include xgboost test)

Why the travis run on scala-2.11 ? I already force it use scala 2.12 in the travis script. @jkbradley

**Comment by WeichenXu123** [comment]
@mengxr PR ready.

File: mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostRegressionModelParitySpec.scala [hunk 3]
```diff
@@ -36,7 +36,7 @@ class XGBoostRegressionModelParitySpec extends FunSpec
   private val xgboostParams: Map[String, Any] = Map(
     "eta" -> 0.3,
     "max_depth" -> 2,
-    "objective" -> "reg:linear",
+    "objective" -> "reg:squarederror",
```
Comment (jkbradley) [inline mleap-xgboost-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/mleap/XGBoostRegressionModelParitySpec.scala]: Note to self: This was an xgboost API change.

File: project/Dependencies.scala [hunk 4]
```diff
@@ -67,8 +67,8 @@ object Dependencies {
       "ch.qos.logback" % "logback-classic" % logbackVersion,
       "com.typesafe.scala-logging" %% "scala-logging" % loggingVersion
     )
-    val xgboostDep = "ml.dmlc" % "xgboost4j" % xgboostVersion
-    val xgboostSparkDep = "ml.dmlc" % "xgboost4j-spark" % xgboostVersion
+    val xgboostDep = "ml.dmlc" % "xgboost4j" % xgboostVersion // scala 2.11 only
```
Comment (jkbradley) [inline project/Dependencies.scala]: For the record: xgboost master is 2.12 only: https://github.com/dmlc/xgboost/pull/4574

**Comment by WeichenXu123** [comment]
@jkbradley @mengxr  The test failed because of sbt fetch latest spark SNAPSHOT version jars from https://repository.apache.org/snapshots/ failed (for scala 2.11). (see log in CI).
 I will ask build team to confirm this.

**Comment by WeichenXu123** [comment]
@mengxr 
When I remove crossScalaVersion, now sbt use scala-2.12 to build, but the weird error raise again:
```
[info] XGBoostRegressionModelParitySpec:
[info] ml.dmlc.xgboost4j.scala.spark.mleap.XGBoostRegressionModelParitySpec *** ABORTED ***
[info]   java.lang.NoClassDefFoundError: scala/Product$class
[info]   at ml.dmlc.xgboost4j.scala.spark.TrackerConf.<init>(XGBoost.scala:52)
[info]   at ml.dmlc.xgboost4j.scala.spark.TrackerConf$.apply(XGBoost.scala:55)
[info]   at ml.dmlc.xgboost4j.scala.spark.params.GeneralParams$class.$init$(GeneralParams.scala:177)
[info]   at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressor.<init>(XGBoostRegressor.scala:51)
[info]   at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressor.<init>(XGBoostRegressor.scala:58)
[info]   at ml.dmlc.xgboost4j.scala.spark.mleap.XGBoostRegressionModelParitySpec.<init>(XGBoostRegressionModelParitySpec.scala:57)
```

File: project/Dependencies.scala (hunk 4 above)
Comment (WeichenXu123) [reply project/Dependencies.scala]: @mengxr Is the comment `// scala 2.11 only` used here as a hint to tell SBT only use scala-2.11 when compiling with this dependency ?

**Comment by ancasarb** [comment]
closing this in favor of https://github.com/combust/mleap/pull/665, thank you!

**Pull request closed without merging**